pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -1155,16 +1156,24 @@ protected <V> MessageImpl<V> newMessage(final MessageIdImpl messageId,
     }
 
     private void executeNotifyCallback(final MessageImpl<T> message) {
+        internalPinnedExecutor.execute(() -> {
+            executeNotifyCallback0(message);
+        });
+    }
+
+    @VisibleForTesting
+    void executeNotifyCallback0(final MessageImpl<T> message) {
```

Comment:
Instead of exposing this internal method for a test, are you able to create a funcational test that verifies this edge case?